### PR TITLE
Add gcc13 to testing matrix.

### PR DESCRIPTION
**Pull request by wmaxey**
## Description

Adds gcc13 to our testing matrix.

## Checklist

- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by gevtushenko** [comment]
Might be blocked by https://github.com/NVIDIA/cccl/issues/1174

File: ci/matrix.yaml [hunk 1]
```diff
@@ -18,8 +18,9 @@ gcc9: &gcc9 { name: 'gcc', version: '9', exe: 'g++' }
 gcc10: &gcc10 { name: 'gcc', version: '10', exe: 'g++' }
 gcc11: &gcc11 { name: 'gcc', version: '11', exe: 'g++' }
 gcc12: &gcc12 { name: 'gcc', version: '12', exe: 'g++' }
+gcc12: &gcc13 { name: 'gcc', version: '13', exe: 'g++' }
```
Comment (jrhemstad) [inline ci/matrix.yaml:21]:
```suggestion
gcc13: &gcc13 { name: 'gcc', version: '13', exe: 'g++' }
```

**Comment by jrhemstad** [comment]
The gcc13 jobs will continue to fail because CUDA 12.3 doesn't officially support gcc13 yet: https://github.com/NVIDIA/cccl/actions/runs/7559215520/job/20582713337?pr=1167#step:8:233

We'd need to add `-allow-unsupported-compiler` to the build flags and decide if that's something we want to just add to the default build flags, or only add it only for the relevant config (gcc13 in this case).

**Comment by miscco** [comment]
Already done

**Pull request closed without merging**